rootex/framework/systems/transform_system.cpp

```diff
@@ -0,0 +1,87 @@
+#include "transform_system.h"
+
+#include "render_system.h"
+
+#define LINE_MAX_VERTEX_COUNT 1000
+#define LINE_VERTEX_COUNT 2
+
+TransformSystem* TransformSystem::GetSingleton()
+{
+	static TransformSystem singleton;
+	return &singleton;
+}
+
+TransformSystem::TransformSystem()
+    : System("TransformSystem", UpdateOrder::Async, true)
+{
+}
+
+void TransformSystem::calculateTransforms(Scene* scene)
+{
+	Entity& entity = scene->getEntity();
+	if (TransformComponent* transform = entity.getComponent<TransformComponent>())
+	{
+		int passDown = transform->getPassDowns();
+		if (SpringArmComponent* springArm = entity.getComponent<SpringArmComponent>())
+		{
+			Vector3 parentPosition = transform->getParentAbsoluteTransform().Translation();
+			Vector3 childPosition = transform->getPosition();
+			Vector3 intialPosition = parentPosition + childPosition;
+			Vector3 finalPosition = parentPosition + springArm->getDesiredLocalPosition();
+
+			transform->setAbsolutePosition(Vector3::Lerp(intialPosition, finalPosition, springArm->getLerp()));
+		}
+		if (passDown == (int)TransformPassDown::All)
+		{
+			pushMatrix(transform->getLocalTransform());
+		}
+		else
+		{
+			Matrix matrix = Matrix::Identity;
+			if (passDown & (int)TransformPassDown::Position)
+			{
+				matrix = Matrix::CreateTranslation(transform->getPosition()) * matrix;
+			}
+			if (passDown & (int)TransformPassDown::Rotation)
+			{
+				matrix = Matrix::CreateFromQuaternion(transform->getRotation()) * matrix;
+			}
+			if (passDown & (int)TransformPassDown::Scale)
+			{
+				matrix = Matrix::CreateScale(transform->getScale()) * matrix;
+			}
+			pushMatrix(matrix);
+		}
+	}
+	else
+	{
+		pushMatrix(Matrix::Identity);
+	}
+
+	for (auto& child : scene->getChildren())
+	{
+		Entity& childEntity = child->getEntity();
+		if (TransformComponent* childTransform = childEntity.getComponent<TransformComponent>())
+		{
+			childTransform->setParentAbsoluteTransform(RenderSystem::GetSingleton()->getCurrentMatrix());
+		}
+
+		calculateTransforms(child.get());
+	}
+	popMatrix();
+}
+
+void TransformSystem::pushMatrix(const Matrix& transform)
+{
+	RenderSystem::GetSingleton()->m_TransformationStack.push_back(transform * RenderSystem::GetSingleton()->m_TransformationStack.back());
```

Comment:
Why make `m_TransformationStack` public? It's not supposed to be accessed like that, and plus, whenever it gets accessed, you need to perform more than just returning the top matrix.